uint256.go

```diff
@@ -369,100 +368,124 @@ func (z *Int) isBitSet(n uint) bool {
-	}
-	if isdividend {
-		n = n + 1
-		dn[n] = prev
+func subMulTo(x, y []uint64, multiplier uint64) uint64 {
+
+	var borrow uint64
+	for i := 0; i < len(y); i++ {
+		s, carry1 := bits.Sub64(x[i], borrow, 0)
+		ph, pl := bits.Mul64(y[i], multiplier)
+		t, carry2 := bits.Sub64(s, pl, 0)
+		x[i] = t
+		borrow = ph + carry1 + carry2
 	}
-	return dn[:n+1]
+	return borrow
 }
 
-func divKnuth(x, y []uint32) []uint32 {
-	m, n := len(x)-1, len(y)
-	q := make([]uint32, m-n+1)
-	// Number base (2**32)
-	var b uint64 = 0x100000000
+func udivremBy1(u []uint64, d uint64) (quot []uint64, rem uint64) {
+	quot = make([]uint64, len(u)-1)
+	rem = u[len(u)-1] // Set the top word as remainder.
 
-	// Take care of the case of a single-digit.
-	if n == 1 {
-		var k uint64
-		k = uint64(x[m])
-		for i := m - 1; i >= 0; i-- {
-			q[i] = uint32((k*b + uint64(x[i])) / uint64(y[0]))
-			k = k*b + uint64(x[i]) - uint64(q[i])*uint64(y[0])
-		}
-		return q
+	for j := len(u) - 2; j >= 0; j-- {
+		quot[j], rem = bits.Div64(rem, u[j], d)
 	}
 
-	// Main Loop
-	var qhat, rhat uint64
-	for j := m - n; j >= 0; j-- {
-		qhat = (uint64(x[j+n])*b + uint64(x[j+n-1])) / uint64(y[n-1])
-		rhat = uint64(x[j+n])*b + uint64(x[j+n-1]) - qhat*uint64(y[n-1])
+	return quot, rem
+}
 
-	AGAIN:
-		if qhat >= b || (qhat*uint64(y[n-2]) > b*rhat+uint64(x[j+n-2])) {
-			qhat = qhat - 1
-			rhat = rhat + uint64(y[n-1])
-			if rhat < b {
-				goto AGAIN
+func udivremKnuth(u, d []uint64) (quot []uint64) {
+	quot = make([]uint64, len(u)-len(d))
+	dh := d[len(d)-1]
+	dl := d[len(d)-2]
+
+	for j := len(u) - len(d) - 1; j >= 0; j-- {
+		u2 := u[j+len(d)]
+		u1 := u[j+len(d)-1]
+		u0 := u[j+len(d)-2]
+
+		var qhat, rhat uint64
+		if u2 >= dh { // Division overflows.
+			qhat = ^uint64(0)
+			// TODO: Add "qhat one to big" adjustment (not needed for correctness, but helps avoiding "add back" case).
+		} else {
+			qhat, rhat = bits.Div64(u2, u1, dh)
+			ph, pl := bits.Mul64(qhat, dl)
+			if ph > rhat || (ph == rhat && pl > u0) {
+				qhat--
+				// TODO: Add "qhat one to big" adjustment (not needed for correctness, but helps avoiding "add back" case).
 			}
 		}
 
 		// Multiply and subtract.
-		var p uint64
-		var t, k int64
-		for i := 0; i < n; i++ {
-			p = qhat * uint64(y[i])
-			t = int64(x[i+j]) - k - int64(p&0xffffffff)
-			x[i+j] = uint32(t)
-			k = int64(p>>32) - (t >> 32)
+		borrow := subMulTo(u[j:], d, qhat)
+		u[j+len(d)] = u2 - borrow
+		if u2 < borrow { // Too much subtracted, add back.
+			qhat--
+			u[j+len(d)] += addTo(u[j:], d)
 		}
-		t = int64(x[j+n]) - k
-		x[j+n] = uint32(t)
-
-		q[j] = uint32(qhat)
-		if t < 0 {
-			// If we subtracted too much, add back.
-			q[j] = q[j] - 1
-			var k, t uint64
-			for i := 0; i < n; i++ {
-				t = uint64(x[i+j]) + uint64(y[i]) + k
-				x[i+j] = uint32(t)
-				k = t >> 32
-			}
-			x[j+n] = x[j+n] + uint32(k)
+
+		quot[j] = qhat // Store quotient digit.
+	}
+	return quot
+}
+
+func udivrem(u []uint64, d *Int) (quot []uint64, rem *Int) {
```

Comment:
This is still implementing `// See Knuth, Volume 2, section 4.3.1, Algorithm D.`, right? please make a docstring